### PR TITLE
Add Karpenter disruption budget example blueprint

### PR DESCRIPTION
**Pull request by RobertNorthard**
*Issue #, if available:*
N/A
*Description of changes:*
Add Karpenter disruption budget example blueprint

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: blueprints/disruption-budgets/README.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
+# Karpenter Blueprint: NodePool Disruption Budgets
+
+## Purpose
+
+A Karpenter’s NodePool disruption budgets can be used to rate limit Karpenter’s voluntary disruptions, if no budget is set their is a default of 10% set. When calculating if a budget will block nodes from disruption, Karpenter checks if the number of nodes being deleted is greater than the number of allowed disruptions. Budgets take into consideration voluntary disruptions through expiration, drift, emptiness and consolidation. If there are multiple budgets defined in the NodePool, Karpenter will honour the most restrictive of the budgets.
```
Comment (ruecarlo) [inline blueprints/disruption-budgets/README.md:5]: I think we may require an entry to explain why disruption controls are needed and be prescriptive on the use case. I'd suggest something like:

"Actions like consolidation, drift detection and ExpireAfter, allow users to optimize for cost in the case of consolidation, keep up with the latest security patches and gold configuration, or ensure governance best practices, like refreshing instances every N days.  This actions cause, as a trade-off, some level of disruption in the cluster caused by expected causes.  To control the trade-off between, for example, being on the latest AMI (drift detection) and nodes restarting when that happen we can use disruption controls and set up disruption budgets. Disruption budgets can help us set a maintenance window in the example above to allow for drift to happen at low use hours for example.

File: blueprints/disruption-budgets/README.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+By applying a combination of disruptions budgets and Pod Disruptions Budgets (PDBs) you get both application and platform voluntary disruption controls, this can help you move towards continually operations to protect workload availability. You can learn more about Karpenter NodePool disruption budges and how the Kapenter disruption controllers works in the [Karpenter documentation](https://karpenter.sh/docs/concepts/disruption/#disruption-controller).
+
+## Examples
+The following provides a set of example disruption budgets:
+
+### Only interrupt a % of Karpenter managed nodes
+
+The following Disruption Budgets says, at any-point in time only disrupt 20% of the Nodes managed by the NodePool. For instance, if there were 19 nodes owned by the NodePool, 4 disruptions would be allowed, rounding up from 19 * .2 = 3.8.
```
Comment (ruecarlo) [inline blueprints/disruption-budgets/README.md:14]: Ideally we should not only explain what is the functionality, but why you would apply it and what are the consequences. an idea/example:

> Applications or systems might be more affected planned disruption caused by consolidation. Some application may find consolidation is either too aggressive, while other might be more fault tolerant and would rather optimize for cost. This configuration to only interrupt a percentage of managed nodes allows you to control the level of churn in this case. This also applies for cases like `ExpireAfter`, where there is an intent to refresh nodes

File: blueprints/disruption-budgets/README.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  ...
+  disruption:
+    consolidationPolicy: WhenUnderutilized
+    expireAfter: 720h
+    budgets:
+    - nodes: "20%"
+```
+
+### Do not disrupt between UTC 09:00 and 18:00 every day
+
```
Comment (ruecarlo) [inline blueprints/disruption-budgets/README.md:31]: When and why would you apply this configuration ?

File: blueprints/disruption-budgets/README.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    expireAfter: 720h
+    budgets:
+    - nodes: "0"
+      schedule: "0 9 * * *"
+      duration: 8h
+```
+
+### Allow 20% disruptions during a maintenance window from UTC 22:00 to 2:00, but only 10% disruptions outside of a maintenance window
+
+The following Disruption Budgets says, for a 4 hour timeframe from UTC 22:00 only disrupt 20% of nodes, but during normal operations only disrupt 10% of nodes.
```
Comment (ruecarlo) [inline blueprints/disruption-budgets/README.md:52]: When and why would you apply this configuration ?

File: blueprints/disruption-budgets/README.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  ...
+  disruption:
+    consolidationPolicy: WhenUnderutilized
+    expireAfter: 720h # 30 * 24h = 720h
```
Comment (ruecarlo) [inline blueprints/disruption-budgets/README.md:90]: Change the 30 * 24h with 30 days... easier to read.

File: blueprints/disruption-budgets/README.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    budgets:
+    - nodes: "20%"
+    - nodes: "5"
```
Comment (ruecarlo) [inline blueprints/disruption-budgets/README.md:93]: We should explain that double entry on node added here and how this applies within the same schedule..

File: blueprints/disruption-budgets/README.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+      duration: 10m
+```
+
+## Requirements
+
+* A Kubernetes cluster with Karpenter installed. You can use the blueprint we've used to test this pattern at the cluster folder in the root of this repository.
+
+## Deploy
+
+Let's say you want to control how nodes are upgraded when switching to Bottlerocket via Karpenter Drift, in this example we deploy a disruption budget, that prevents disruptions 24 hours a day 7 days a week. You can use the schedule and duration of the budget to control when disruptions via Drift can take place.
```
Comment (ruecarlo) [inline blueprints/disruption-budgets/README.md:105]: This is perfect 👍

**Review by ruecarlo — APPROVED** [review]
Just left some comments. Great work  !!! Thanks so much for this. This will be great to clarify and build a blog/slides to help provide more prescriptive guidance !

File: blueprints/disruption-budgets/README.md (hunk 2 above)
Comment (RobertNorthard) [reply blueprints/disruption-budgets/README.md:14]: Like it, thank you!

File: blueprints/disruption-budgets/README.md (hunk 3 above)
Comment (RobertNorthard) [reply blueprints/disruption-budgets/README.md:33]: Thank you, added:

`You might apply this configuration if you would not like Karpenter to not disrupt workloads during times when the workload might be receiving peak traffic.`

File: blueprints/disruption-budgets/README.md (hunk 4 above)
Comment (RobertNorthard) [reply blueprints/disruption-budgets/README.md:58]: The use case I see for this is:

`You might apply this configuration if outside of core business hours you prefer a higher set of disruptions to speed up the role out of AMI updates for example.`

File: blueprints/disruption-budgets/README.md (hunk 6 above)
Comment (RobertNorthard) [reply blueprints/disruption-budgets/README.md:101]: Thank you, added that even though we can disrupt 20% of nodes, as both budgets are active karpenter will honor the most restrictive which means only a maximum of 5 nodes at any time.

**Pull request merged**